### PR TITLE
Nos3#98 mag

### PR DESCRIPTION
**Pull request by wmhollan**
Pull request to merge in the generic magnetometer component and associated files.

### CODE REVIEW

**Review by Isgaroth-the-Green — APPROVED** [review]
Everything looks reasonable and, based on the testing I have done of the commands, it all appears to function properly as well.

File: sim/src/generic_mag_data_point.cpp [hunk 1]
```diff
@@ -10,18 +10,22 @@ namespace Nos3
         sim_logger->trace("Generic_magDataPoint::Generic_magDataPoint:  Defined Constructor executed");
```
Comment (msuder) [inline sim/src/generic_mag_data_point.cpp]: Delete this constructor.  In addition, please delete the files sample_data_provider.cpp/.hpp.  This constructor and those two files are unused and unneeded.

File: sim/src/generic_mag_data_point.cpp [hunk 2]
```diff
@@ -41,30 +45,45 @@ namespace Nos3
...
-                    /* Debug print */
-                    sim_logger->trace("Generic_magDataPoint::Generic_magDataPoint:  Parsed svb = %f %f %f", _generic_mag_data[0], _generic_mag_data[1], _generic_mag_data[2]);
+                    size_t lb = lines[i].find_first_of("[", MSsize);
+                    size_t rb = lines[i].find_first_of("]", MSsize);
+                    int index = std::stoi(lines[i].substr(lb+1, rb-lb-1));
+                    if ((index >= 0) && (index < numAxes)) {
+                        std::string param(lines[i].substr(rb+2, 5));
+                        size_t equal = lines[i].find_first_of("=");
+                        std::string value(lines[i].substr(equal+1, lines[i].size()-equal-1));
+                        if (param.compare("Valid") == 0)
```
Comment (msuder) [inline sim/src/generic_mag_data_point.cpp]: I don't believe 42 has a "Valid" flag for magnetometers.  Remove lines 61-71 (and change the beginning of line 72 from "} else if" to "if").  Also remove the variable declaration of "valid" on line 26.

File: README.md [hunk 3]
```diff
@@ -3,55 +3,30 @@ This repository contains the NOS3 Generic_mag Component.
 This includes flight software (FSW), ground software (GSW), simulation, and support directories.
 
 ## Overview
-This generic_mag component is a UART device that accepts multiple commands, including requests for telemetry and data.
-The available FSW is for use in the core Flight System (cFS) while the GSW supports COSMOS.
+This generic_mag component is an SPI device that accepts multiple commands including requests for telemetry. 
+The available FSW is for use in the core Flight System (cFS) while the GSw supports COSMOS.
 A NOS3 simulation is available which includes both generic_mag and 42 data providers.
```
Comment (msuder) [inline README.md]: Change "both generic_mag and 42 data providers" to "a 42 data provider".  I put in a comment later to remove the generic_mag_data_provider.cpp/.hpp files and one of the Generic_magDataPoint constructors.

File: README.md [hunk 4]
```diff
@@ -62,7 +37,7 @@ Refer to the file [fsw/platform_inc/generic_mag_platform_cfg.h](fsw/platform_inc
 configuration settings, as well as a summary on overriding parameters in mission-specific repositories.
 
 ## Simulation
-The default configuration returns data that is X * 0.001, Y * 0.002, and Z * 0.003 the request count after conversions:
+The default configuration returns data that is X * 0.001, Y * 0.001, and Z * 0.001 the request count after conversions:
```
Comment (msuder) [inline README.md]: Delete this line since we are deleting the generic data provider.

File: fsw/src/generic_mag_events.h [hunk 5]
```diff
@@ -43,10 +43,7 @@
 #define GENERIC_MAG_REQ_DATA_ERR_EID          32
 
 /* Hardware protocol event IDs */
-#define GENERIC_MAG_UART_INIT_ERR_EID         40
-#define GENERIC_MAG_UART_CLOSE_ERR_EID        41
-#define GENERIC_MAG_UART_READ_ERR_EID         42
-#define GENERIC_MAG_UART_WRITE_ERR_EID        43
-#define GENERIC_MAG_UART_TIMEOUT_ERR_EID      44
+#define GENERIC_MAG_SPI_INIT_ERR_EID          40
+#define GENERIC_MAG_SPI_READ_ERR_EID          41
```
Comment (msuder) [inline fsw/src/generic_mag_events.h:41]: Do not appear tp be used:
GENERIC_MAG_REQ_HK_ERR_EID
GENERIC_MAG_CONFIG_ERR_EID
GENERIC_MAG_REQ_HK_ERR_EID

File: sim/CMakeLists.txt [hunk 6]
```diff
@@ -1,7 +1,7 @@
 project(generic_mag_sim)
 
 find_package(ITC_Common REQUIRED QUIET COMPONENTS itc_logger)
-find_package(NOSENGINE REQUIRED QUIET COMPONENTS common transport client uart)
+find_package(NOSENGINE REQUIRED QUIET COMPONENTS common transport client spi)
 
 include_directories(inc
                     ${sim_common_SOURCE_DIR}/inc
```
Comment (msuder) [inline sim/CMakeLists.txt:7]: Delete generic_mag_data_provider.cpp on line 14 below.

File: sim/src/generic_mag_data_point.cpp [hunk 7]
```diff
@@ -41,30 +45,45 @@ namespace Nos3
...
+                        }
+                    }
                 }
             }
         } 
         catch(const std::exception& e) 
         {
-            /* Report error */
+            /* Force data to be set to known values */
+            std::vector<float> axes(3, 5.0);
```
Comment (msuder) [inline sim/src/generic_mag_data_point.cpp]: I think change "5.0" to "0.0" as the known value.

File: sim/src/generic_mag_data_point.cpp [hunk 8]
```diff
@@ -73,20 +92,15 @@ namespace Nos3
...
-        ss << " Generic_mag Data: "
-           << _generic_mag_data[0]
-           << " "
-           << _generic_mag_data[1]
-           << " "
-           << _generic_mag_data[2];
-
-        return ss.str();
+        std::stringstream output;
+        output << "Channel values: ";
```
Comment (msuder) [inline sim/src/generic_mag_data_point.cpp]: Change "Channel" to "Magnetometer".

**Review by msuder — APPROVED** [review]

**Review by jlucas9 — APPROVED** [review]
Confirmed fully functional in local testing

**Pull request merged**